### PR TITLE
wallet: Fix -maxtxfee check by moving it to CWallet::CreateTransaction

### PR DESCRIPTION
**Pull request by promag**
Follow up to #16257, this PR makes `bumpfee` aware of `-maxtxfee`.

It also prevents dangling locked unspents when calling `fundrawtransaction` - because the previous check was after `LockCoin`.

### CODE REVIEW

**Comment by promag** [comment]
@Sjors you promised a review 😄

**Comment by Sjors** [comment]
tACK a2f23b8

This also makes the QT message the same as RPC.

Afaik there's no way to reach the `A fee higher than %1 is considered an absurdly high fee` message anymore, so should we get rid of that string and `Reject absurdly high fee` in `walletmodel.cpp`?

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -3135,6 +3130,11 @@ bool CWallet::CreateTransaction(interfaces::Chain::Lock& locked_chain, const std
         }
     }
 
+    if (nFeeRet > this->m_default_max_tx_fee) {
```
Comment (MarcoFalke) [inline src/wallet/wallet.cpp]: style-nit:

```suggestion
    if (nFeeRet > m_default_max_tx_fee) {
```

`m_` already means `this->`, no need for duplicate.

File: src/wallet/wallet.cpp [hunk 2]
```diff
@@ -3135,6 +3130,11 @@ bool CWallet::CreateTransaction(interfaces::Chain::Lock& locked_chain, const std
...
+        strFailReason = TransactionErrorString(TransactionError::MAX_FEE_EXCEEDED);
```
Comment (MarcoFalke) [inline src/wallet/wallet.cpp]: style-nit: Maybe for the purpose of a backport, this could be inlined as a string literal?

**Review by maflcko — COMMENTED** [review]
Concept ACK.

Maybe revert the docstring in the gui to say "belt-and-suspenders" again?

See https://github.com/bitcoin/bitcoin/pull/16257/files#diff-2e3836af182cfb375329c3463ffd91f8

**Comment by MarcoFalke** [comment]
Looks good, I'll write some tests and send them over in a bit.

File: src/wallet/wallet.cpp (hunk 2 above)
Comment (promag) [reply src/wallet/wallet.cpp]: Don't you prefer an unclean-but-easy-cherry-pick backport?

**Comment by MarcoFalke** [comment]
ACK, tested by writing a test: https://github.com/MarcoFalke/bitcoin/commit/9ff66a1b2157bcce8ccd4cea0610eae79d0be6d8

**Comment by Sjors** [comment]
Thanks for the `sendmany ` and `fundrawtransaction` RPC tests @MarcoFalke.

> Maybe revert the docstring in the gui to say "belt-and-suspenders" again?

Maybe that is indeed slightly safer than dropping it completely.

**Comment by promag** [comment]
@MarcoFalke great stuff, pulled both commits.

**Comment by MarcoFalke** [comment]
ACK 9ff66a1b2157bcce8ccd4cea0610eae79d0be6d8 (read code, wrote test case)

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 9ff66a1b2157bcce8ccd4cea0610eae79d0be6d8 (read code, wrote test case)
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUhCrQv8C829o7ZZs8hUWaldWPLHQdivLOKMhyEC24GVjlOb2CPwh+YDSs4bZUNj
qVpgveclR55eb9j4tj7Wt+4AqNiNeHyKMgNfXCXeWKj4CYDSGRfwd1zpSnL8Z/Lk
cXyp2AV5GcSX3hjG/1tbFphxDYLoiihfSLMVTGybC4aNupxqIiAFRPBVS9ovwHBA
Ml3r9H0CueguuZMR1m2B/hxA+fOWe19VxCck93dydfL7v+N56tvatm/+5ihwaXAf
v3wwrHHdA2j0i3yXOSCdDezC0arwA3CMVJWZNt5rJt2ydLG46YSp7S4XwjpJD36V
G04nwXgfIRv9HPPeYLESOQHf/RiGQaaC4TRLuxPo3P/abhzXq2ucSTdsm0mtqeuj
pTdasa8y6fAWs4Y//y84ce28bbLpWwM1QHCbf1GwemqNhXpxOwZnk6uAqwM+2D6v
SJ0hV5HqocIfbUM1WKqBnKpLnvHicvPeMOjTtQw3g1KHvhY9DCK9tAX1vYom2qDd
Y5xHfFCm
=I+BI
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `e8c4afba47440f5e82e36c70bed731c61cc784c18d70774903ee95050752dbfa  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108e8c4afba47440f5e82e36c70bed731c61cc784c18d70774903ee95050752dbfaf010459a675ee30854e747ebfe5932880d8108fff01040420c478dd0857f92485b2465b9e3f308f020e0241732d39ce0372f70690e10902da5d17e88634706fed07f2e750b598dc04608f1045d1bbbb6f00829b7303c8610f4b60083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff01046353fe8e34a01509e32b2d3865a17a108f1045d1bbbb5f0084bc64dfb500b061e0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010d840b2963253afa7ad168a848461f4d408f1045d1bbbb5f0083827364461e125be0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6df010f397c9f8bd1fc17abf6953a5c175e92708f1045d1bbbb5f00863fbbec4f0fc0a4e0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6d)

</details>

**Comment by laanwj** [comment]
Concept ACK

**Comment by promag** [comment]
@meshcollider ping.

**Comment by MarcoFalke** [comment]
The previous pull was reviewed by @kallewoof @jonasschnelli @laanwj @achow101. Would be helpful to get your eyes on this as well.

**Review by kallewoof — APPROVED** [review]
utACK 9ff66a1b2157bcce8ccd4cea0610eae79d0be6d8

Looks good. Was wondering about the two test commits, until I realized author was different.

File: test/functional/wallet_create_tx.py [hunk 3]
```diff
@@ -38,6 +43,40 @@ def run_test(self):
...
+                "Fee exceeds maximum configured by -maxtxfee",
+                lambda: self.nodes[0].sendmany(dummy="", amounts=outputs),
+            )
+            assert_raises_rpc_error(
+                -4,
+                "Fee exceeds maximum configured by -maxtxfee",
+                lambda: self.nodes[0].fundrawtransaction(hexstring=raw_tx),
+            )
+
+        self.log.info('Check maxtxfee in combination with settxfee'.format(fee_setting))
```
Comment (practicalswift) [inline test/functional/wallet_create_tx.py]: `.format(fee_setting)` part is a mistake? :-)

Comment (promag) [reply test/functional/wallet_create_tx.py]: Yao, @MarcoFalke I'll amend.

**Comment by MarcoFalke** [comment]
re-ACK 0d101a340c44841cbbc5982d55354b1787bc39e2, only change is small test fixup

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK 0d101a340c44841cbbc5982d55354b1787bc39e2, only change is small test fixup
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUgQbwwAtg+GMzWeDLh1+NCCGz5Bs6tmgzthA2R28KiIaoQsgivJMdd2FoOMl7/C
dlKZ7gKbttnEtr/Rl4pFALzy9L8oMiFmoXC+qy3APx4Z3Xd+MS4aLM1WqxHggkPN
u6679W5/OtN4UE6oksKGM84EOAfKqSCHG/cZqGJ8a4UCIZWku2EhHiqo16gaNJvi
5MNuaIQjpGqQf5egbqV+lWIvJmnHXzoIKN09Eo7PS7PwtCJy6A1x8uy12CxsU6G3
Dl8M27a52PNvoXHeWX+xyKlbDIFSOf9vuc4/bP/ylMtC7vIGdO/W0SIyvsaNMLHq
+cOAJZxLoxZKgK2VUQqxH1L/VafPtm/a2v23LzkUFoK7Vk0fENekpuiQTZsDRQV8
OOXhf7TFukNGVCS60tBJa1Mf0LrSfT3DgV/NNVni1YYDg97H6tX0lpXQyJ8LSZg5
o8dWBno9MWsoFAyAfsPxWtS9FI7xLb1XuYMiMNBdF2FGikHIebt9R5CJqfGVIgoc
9IJFNrPA
=Cuz4
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `a2d0938771647864e65f9d3c8939d69334e4a3be00ac928e24814e25e15140d2  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108a2d0938771647864e65f9d3c8939d69334e4a3be00ac928e24814e25e15140d2f010f34d968b25b41f51b37a69b589dad14108fff01022f5741753928c82bbcc2e8f013eb01708f1045d25d26cf0082924763748fee4aa0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff010249906d5a5197a98b77bd77458daab0408f1045d25d26cf0084d8d65fb09f2ad980083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff01067af5990bbcc75c69c6a00b30a0e02aa08f1045d25d26cf008add2d6d30c0e5dcb0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010e4c568e21c0c9ce5d6c41a74b5bba52b08f1045d25d26cf008d5df6cab1351f3180083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

**Comment by laanwj** [comment]
ACK 0d101a340c44841cbbc5982d55354b1787bc39e2

**Comment by fanquake** [comment]
Being backported in #16414.

**Pull request merged**